File: bindings/java-jni/src/main/java/org/maplibre/nativejni/resource/ResourceTransformCallback.java
package org.maplibre.nativejni.resource;

import java.util.Optional;

/**
 * Rewrites network resource URLs for a runtime.
 *
 * <p>Native code may invoke this callback on worker or network threads. The callback should return
 * quickly and avoid calling Maplibre APIs. Returning {@link Optional#empty()} keeps the original
 * URL. The binding catches callback exceptions and treats them as no rewrite.
 */
@FunctionalInterface
public interface ResourceTransformCallback {
  Optional<String> transform(ResourceTransformRequest request);
}

File: bindings/kotlin/src/commonMain/kotlin/org/maplibre/nativeffi/resource/ResourceTransformCallback.kt
package org.maplibre.nativeffi.resource

/**
 * Rewrites network resource URLs for a runtime.
 *
 * Native code may invoke this callback on worker or network threads. The callback should return
 * quickly and avoid calling Maplibre APIs. Returning null keeps the original URL. The binding
 * catches callback exceptions and treats them as no rewrite.
 */
public fun interface ResourceTransformCallback {
  public fun transform(request: ResourceTransformRequest): String?
}
